src/app/services/user-deposit.service.ts

```diff
@@ -0,0 +1,124 @@
+import { Contract } from 'web3-eth-contract';
+
+@Injectable({
+    providedIn: 'root',
+})
+export class UserDepositService {
+    readonly balance$: Observable<BigNumber>;
+    readonly servicesToken$: Observable<UserToken>;
+
+    private servicesToken: UserToken;
+
+    constructor(
+        private raidenService: RaidenService,
+        private tokenInfoRetriever: TokenInfoRetrieverService,
+        private raidenConfig: RaidenConfig
+    ) {
+        const userDepositContract$ = this.raidenService.reconnected$.pipe(
+            switchMap(() => this.raidenService.getContractsInfo()),
+            map(
+                (contracts) =>
+                    new this.raidenConfig.web3.eth.Contract(
+                        userDepositAbi,
+                        contracts.user_deposit_address
+                    )
+            ),
+            backoff(this.raidenConfig.config.error_poll_interval),
+            shareReplay(1)
+        );
+
+        this.balance$ = this.getBalanceObservable(userDepositContract$);
+        this.servicesToken$ = this.getServicesTokenObservable(
+            userDepositContract$
+        );
+    }
+
+    private getBalanceObservable(
+        userDepositContract$: Observable<Contract>
+    ): Observable<BigNumber> {
+        const fetchBalance$ = zip(
+            this.raidenService.raidenAddress$,
+            userDepositContract$
+        ).pipe(
+            switchMap(([raidenAddress, contract]) =>
+                fromPromise<string>(
+                    contract.methods.balances(raidenAddress).call()
+                )
+            ),
+            map((value) => new BigNumber(value)),
+            retryWhen((errors) =>
+                errors.pipe(switchMapTo(this.raidenService.rpcConnected$))
+            )
+        );
+
+        return interval(15000).pipe(
+            startWith(0),
+            switchMapTo(fetchBalance$),
+            shareReplay({ bufferSize: 1, refCount: true })
+        );
+    }
+
+    private getServicesTokenObservable(
+        userDepositContract$: Observable<Contract>
+    ): Observable<UserToken> {
+        const servicesTokenAddress$ = userDepositContract$.pipe(
+            switchMap((contract) =>
+                fromPromise<string>(contract.methods.token().call())
+            ),
+            tap(() => {
+                this.servicesToken = undefined;
+            }),
+            retryWhen((errors) =>
+                errors.pipe(switchMapTo(this.raidenService.rpcConnected$))
+            ),
+            shareReplay(1)
+        );
+
+        const fetchTokenInfo$ = zip(
+            this.raidenService.raidenAddress$,
+            servicesTokenAddress$
+        ).pipe(
+            switchMap(([raidenAddress, tokenAddress]) =>
+                fromPromise(
+                    this.tokenInfoRetriever.createBatch(
+                        [tokenAddress],
+                        raidenAddress,
+                        { [tokenAddress]: this.servicesToken }
```

Comment:
It looks a bit weird, I know. This object is needed for the TokenInfoRetriever to only fetch the static token properties (name, decimals, symbol) once. So later only the balance is fetched. It is right that `servicesToken` should be `undefined` the first time `createBatch` is called.

The `serviceToken` property should be unset whenever a new userDepositContract is emitted (after Raiden restarted). So afterwards all the properties get fetched once again.

Reply:
Hmm. So it is more a temporally thing.
I must admit that this is a problem for because it makes the code harder to understand. If would have to refactor it, I would first need to fully understand this awkward (sorry for the term) behavior to be sure I don't introduce a bug.
But I must also admit I fail to come up with a better solution right now. So I would close this conversation if you are at least sure that any test case would fetch it if somebody in future mix this up. Then it is at least secured that he must somehow understand it to implement it right (again) and the bug does not get introduced unnoticed.

Reply:
Actually I think now the `createBatch` method should only get called once. It does not really make sense to send a batch request only for the `balanceOf` method. The `balanceOf` method could get called directly every interval.
And that way we don't need to store the `servicesToken`. I hope that makes it clearer.

Reply:
Sounds much better. Could that it helped to simply rethink it again. :upside_down_face:

Reply:
Yep thank you!